### ISSUE TITLE
native cast principal to string and vice versa

### ISSUE THREAD

**Issue by CyrusVorwald** [clarity-lang/reference#74]

**Comment by obycode** [comment]
See #82 for proposal for a generic `to-string` builtin which would include principals. Maybe a `from-string?` would also make sense, similar to `from-consensus-buff?`.

**Comment by jbencin** [comment]
I think it makes more sense to have `*-from?` functions rather than `from-*?` functions. For example, if you write:

```clarity
(let ((x (from-string? "100")))
  (print x)
)
```

Then what type is `x`? Is it an `int` or a `uint`? If you do this instead:

```clarity
(let ((x (uint-from? "100")))
  (print x)
)
```

there is never any ambiguity

**Comment by jbencin** [comment]
> Would the *-from? then support multiple input types?

Yes, this seems much simpler than the other option of having multiple *output* types on a function like `from-string`.

There are several issues on this already, so I guess it's agreed that we need generic type conversion functions. The remaining question is how do we name them?

- `to-string?`
- `string-from?`

These are semantically equivalent, they both take any type and return a string if the conversion is successful, but maybe we should stick with `to-<type>` syntax because it matches  our existing function naming better (and they would all appear together in the docs, which are alphabetically sorted)

So to summarize, here's what I'm thinking so far:
- `to-int` and `to-uint` already exist but should take `bool`, `string-*`, and `buff` types (`buff` requires additional arg for endianness)
- Add `to-buff` for all primitive types (`int` and `uint` require additional arg for endianness)
- Add `to-bool` for all primitive types except `principal`
- Add `to-principal` for `string-*` types
- Add `to-ascii` and `to-utf8` which should accept any type
- Deprecate the following functions:
  - `string-to-int?` and `string-to-uint?`
  - `int-to-ascii` and `int-to-utf8`
  - `buff-to-uint-be`, `buff-to-uint-le`, `buff-to-int-be`, and `buff-to-int-le`

I think this covers all valid conversions between primitive types. Also, do we want to add a `?` to the conversions that might fail, since these function might fail for some input types but always work for others?
